router/rate_limiting.go

```diff
@@ -0,0 +1,109 @@
+package router
+
+import (
+	"container/list"
+	"context"
+	"sync"
+	"time"
+
+	"github.com/topfreegames/pitaya/cluster"
+	"github.com/topfreegames/pitaya/config"
+	"github.com/topfreegames/pitaya/constants"
+	"github.com/topfreegames/pitaya/errors"
+	"github.com/topfreegames/pitaya/logger"
+	"github.com/topfreegames/pitaya/route"
+	"github.com/topfreegames/pitaya/session"
+)
+
+// WithRateLimiting wraps routingFunc and controls
+// player rate through session.
+func WithRateLimiting(c *config.Config, f RoutingFunc) RoutingFunc {
+	limit := c.GetInt("pitaya.router.ratelimiting.limit")
+	interval := c.GetDuration("pitaya.router.ratelimiting.interval")
+	forceDisable := c.GetBool("pitaya.router.ratelimiting.forcedisable")
+	if forceDisable {
+		logger.Log.Warn("rate limiting is forcibly disabled")
+	}
+
+	return func(
+		ctx context.Context,
+		r *route.Route,
+		payload []byte,
+		servers map[string]*cluster.Server,
+	) (*cluster.Server, error) {
+		if forceDisable {
+			return f(ctx, r, payload, servers)
+		}
+
+		sessionVal := ctx.Value(constants.SessionCtxKey)
+		if sessionVal == nil {
+			logger.Log.Warnf("session not found, cant rate limit, route: %s", r.String())
```

Comment:
"can't"

also, when would this case happen (not having a session)?

Reply:
Just for safety, so the application doesn't crash. But I will verify if this can't happen at all and, if so, I remove this check